### PR TITLE
expression: rewrite builtin function: COMPRESS, UNCOMPRESS, UNCOMPRESSED_LENGTH

### PR DESCRIPTION
**Pull request by breeswish**
Fix https://github.com/pingcap/tidb/issues/4091

PTAL @zz-jason @XuHuaiyu

### CODE REVIEW

**Comment by breeswish** [comment]
#4080

File: expression/builtin_encryption.go [hunk 1]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
 
-// eval evals a builtinUncompressSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompress
-func (b *builtinUncompressSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+type compressFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *compressFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: `verifyArgs(args)` firstly

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
+	}
+
+	return string(buffer), false, nil
+}
+
+type uncompressFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *uncompressFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: ditto

File: expression/builtin_encryption.go [hunk 3]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
+	return string(bytes), false, nil
 }
 
 type uncompressedLengthFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *uncompressedLengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinUncompressedLengthSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: ditto

File: expression/builtin_encryption.go [hunk 4]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
+func (b *builtinUncompressedLengthSig) evalInt(row []types.Datum) (int64, bool, error) {
+	payload, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
 	if err != nil {
-		return d, errors.Trace(err)
+		return 0, true, errors.Trace(err)
 	}
-	arg := args[0]
-	if arg.IsNull() {
-		return d, nil
+	if isNull {
```
Comment (zz-jason) [inline expression/builtin_encryption.go]:
```go
if isNull || err != nil {
    return 0, true, errors.Trace(err)
}
```

File: expression/builtin_encryption.go [hunk 5]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
-	if err != nil {
-		return d, errors.Trace(err)
+	if len(payload) == 0 {
+		return 0, false, nil
 	}
-	if len(compressStr) == 0 {
-		d.SetInt64(0)
-		return d, nil
+	if len(payload) <= 4 {
+		// TODO: Generate warning.
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: generating warning in some particular sql mode ? you can reference the cast function.

File: expression/builtin_encryption.go [hunk 6]
```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
...
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString)
 	if err != nil {
-		return d, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	arg := args[0]
-	if arg.IsNull() {
-		return d, nil
+	srcLen := args[0].GetType().Flen
+	compressBound := srcLen + (srcLen >> 12) + (srcLen >> 14) + (srcLen >> 25) + 13
```
Comment (breeswish) [inline expression/builtin_encryption.go]: MySQL's `flen` seems to be `12+len*3.6`. However I didn't find magic number `3.6` in source code and I'm not sure about this.

I ported compressBound from zlib's `compressBound` function, which gives an upper bound of the length of compressed data: https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/compress.c#L81

**Comment by breeswish** [comment]
@zz-jason Done, PTAL

**Comment by breeswish** [comment]
@zz-jason @XuHuaiyu PTAL

**Comment by zz-jason** [comment]
@breeswish plz resolve conflicts

**Pull request merged**